mobile/src/main/java/org/openhab/habdroid/ui/homescreenwidget/ItemUpdateWidget.kt

```diff
@@ -0,0 +1,237 @@
+                } catch (e: HttpClient.HttpException) {
+                    Log.e(TAG, "Error downloading icon for url $iconUrl", e)
+                } catch (e: IOException) {
+                    Log.e(TAG, "Error saving icon to disk", e)
+                }
+            }
+
+            setIcon(context, views, smallWidget, appWidgetId, appWidgetManager)
+        }
+
+        private fun setIcon(
+            context: Context,
+            views: RemoteViews,
+            smallWidget: Boolean,
+            appWidgetId: Int,
+            appWidgetManager: AppWidgetManager
+        ) = GlobalScope.launch {
+            val bitmap = try {
+                context.openFileInput(getFileNameForWidget(appWidgetId)).use {
+                if (context.getPrefs().isIconFormatPng()) {
+                    val byteArray = it.readBytes()
+                    BitmapFactory.decodeStream(byteArray.inputStream())
+                } else {
+                    val widgetOptions = appWidgetManager.getAppWidgetOptions(appWidgetId)
+                    var height = widgetOptions.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_HEIGHT).toFloat()
+                    val width = widgetOptions.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_WIDTH).toFloat()
+                    if (!smallWidget) {
+                        height *= 0.5F
+                    }
+                    // Image view height is 50% of the widget height
+                    val sizeInDp = min(height, width)
+                    @Px val size = context.resources.dpToPixel(sizeInDp).toInt()
+                    Log.d(TAG, "Icon size: $size")
+                    val svg = String(it.readBytes())
+                    bitmapToSvg(svg, size)
+                }
+            }
+            } catch (e: IOException) {
+                Log.e(TAG, "Error getting icon from disk", e)
+                null
+            }
+
+            if (bitmap == null) {
+                Log.d(TAG, "Bitmap is null")
+                views.setImageViewResource(R.id.item_icon, R.mipmap.icon)
+            } else {
+                views.setImageViewBitmap(R.id.item_icon, bitmap)
+            }
+
+            appWidgetManager.updateAppWidget(appWidgetId, views)
+        }
+
+        fun getInfoForWidget(context: Context, id: Int): ItemUpdateWidgetData? {
+            val prefs = context.getSharedPreferences(getFileNameForWidget(id), MODE_PRIVATE)
+            val item = prefs.getString(PreferencesActivity.ITEM_UPDATE_WIDGET_ITEM)
+            val state = prefs.getString(PreferencesActivity.ITEM_UPDATE_WIDGET_STATE)
+            val label = prefs.getString(PreferencesActivity.ITEM_UPDATE_WIDGET_LABEL)
+            val mappedState = prefs.getString(PreferencesActivity.ITEM_UPDATE_WIDGET_MAPPED_STATE)
+            val icon = prefs.getString(PreferencesActivity.ITEM_UPDATE_WIDGET_ICON)
+            if (item.isEmpty() || state.isEmpty() || label.isEmpty() || mappedState.isEmpty()) {
+                return null
+            }
```

Comment:
This check can be removed, as you won't get broadcasts for non-existing widgets.